src/Coverage/Coverable.hs

```diff
@@ -0,0 +1,76 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE NamedFieldPuns #-}
+module Coverage.Coverable where
+
+import           AST.Solved
+import           Explain.Location
+import qualified Data.Set                      as S
+
+data Coverable
+  = Function { line :: Int, name :: String }
+  | Line { line :: Int }
+  deriving (Ord, Eq, Show)
+
+collectFromAST :: AST -> [Coverable]
+collectFromAST AST { aexps } =
+  S.toList $ S.fromList $ concat $ collect <$> aexps
+
+class Collectable a where
+  collect :: a -> [Coverable]
+
+instance Collectable Exp where
+  collect (Solved _ (Area (Loc _ l _) _) exp) = case exp of
+    Assignment name (Solved _ (Area (Loc _ line _) _) (Abs _ body)) ->
+      [Function { line = line, name = name }, Line { line = line }]
```

Comment:
What is happening here?

Reply:
Here I gather the lcov.info data before the test run. So say, that's a section:

```
TN:
SF:/Users/a.boeglin/Code/madlib/fixtures/Test.mad
FN:3,add  // I gather this, so line 3, there's a function add
FNDA:2,add
FNF:1  // Amount of functions found
FNH:1
DA:3,4
DA:4,8
LF:2  // Amount of lines found that should be covered
LH:2
end_of_record
```

Then everything that is not commented is set to 0 at compile time. So mainly I set the target for the coverage. Then when it runs I gather all "hit points", and update the file with these.